src/FakeItEasy/Creation/DelegateProxies/DelegateProxyGenerator.cs

```diff
@@ -74,24 +74,108 @@ private static Delegate CreateDelegateProxy(Type typeOfProxy, MethodInfo invokeM
         }
 
         private static Expression CreateBodyExpression(
-            MethodInfo delegateMethod, DelegateCallInterceptedEventRaiser eventRaiser, IEnumerable<Expression> parameterExpressions)
+            MethodInfo delegateMethod,
+            DelegateCallInterceptedEventRaiser eventRaiser,
+            ParameterExpression[] parameterExpressions)
         {
-            var parameterExpressionsCastToObject =
-                parameterExpressions.Select(x => Expression.Convert(x, typeof(object))).Cast<Expression>().ToArray();
-
-            Expression body = Expression.Call(
+            /*
+             * Generate a method that:
+             * - wraps its arguments in an object array
+             * - passes this array to eventRaiser.Raise()
+             * - assigns the output values back to the ref/out parameters
+             * - casts and returns the result of eventRaiser.Raise()
+             *
+             * For instance, for a delegate like this:
+             *
+             * delegate int Foo(int x, ref int y, out int z);
+             *
+             * We generate a method like this:
+             *
+             * int ProxyFoo(int x, ref int y, out int z)
+             * {
+             *     var arguments = new[]{ (object)x, (object)y, (object)z };
+             *     var result = (int)eventRaiser.Raise(arguments);
+             *     y = (int)arguments[1];
+             *     z = (int)arguments[2];
+             *     return result;
+             * }
+             *
+             * Or, for a delegate with void return type:
+             *
+             * delegate void Foo(int x, ref int y, out int z);
+             *
+             * void ProxyFoo(int x, ref int y, out int z)
+             * {
+             *     var arguments = new[]{ (object)x, (object)y, (object)z };
+             *     eventRaiser.Raise(arguments);
+             *     y = (int)arguments[1];
+             *     z = (int)arguments[2];
+             * }
+             */

```

Comment:
I'm late to the game here, but I find this comment style a little odd. The leading `*` on each line seems redundant. I would expect

```C#
            /*
            Generate a method that:
            - wraps its arguments in an object array
            - passes this array to eventRaiser.Raise()
            ...
            */
```

or better still

```C#
            // Generate a method that:
            // - wraps its arguments in an object array
            // - passes this array to eventRaiser.Raise()
            ...
```

This would be consistent with other places where we have multi-line comments (I've never been a fan of `/* ... */` style comments):

* https://github.com/FakeItEasy/FakeItEasy/blob/3724e900b687981898f91d17c87c742568b5fce8/src/FakeItEasy/Core/FakeManager.EventRule.cs#L58-L62
* https://github.com/FakeItEasy/FakeItEasy/blob/3724e900b687981898f91d17c87c742568b5fce8/tests/FakeItEasy.Tests/Core/ArgumentConstraintTrapTests.cs#L68-L76

There are also quite a few 3 line comments and even plenty of 2 line comments, and they all use the `//` style.